0004-metric-configurable-aggregation.md

```diff
@@ -0,0 +1,59 @@
+1. Indicating something about the default interpretation, based on the action verb (Set, Inc, Record, etc.)
+  1. For Gauges: the action is Set()
+  1. For Cumulatives: the action is Inc()
+  1. For Measures: the action is Record()
+1. Unless the programmer declares otherwise, suggesting a default aggregation
+  1. For Gauges: LAST_VALUE is interesting, SUM is likely not interesting
+  1. For Cumulatives: SUM is interesting, LAST_VALUE is likely not interesting
+  1. For Measures: all aggregations apply, default is MIN, MAX, SUM, COUNT.
+
+## Internal details
+
+Metric constructors should take an optional list of aggregations, to override the default behavior.  When constructed with an explicit list of aggregations, the implementation may use this as a hint about which aggregations should be exported by default.  However, the implementation is not bound by these recommendations in any way and is free to control which aggregations that are applied.
+
+The standard defined aggregations are broken into two groups, those which are "decomposable" (i.e., inexpensive) and those which are not.
+
+The decomposable aggregations are simple to define:
+
+1. SUM: The sum of observed values.
+1. COUNT: The number of observations.
+1. MIN: The smallest value.
+1. MAX: The largest value.
+1. LAST_VALUE: The latest value.
+
+The non-decomposable aggregations do not have standard definitions, they are purely advisory.  The intention behind these are:
+
+1. HISTOGRAM: The intended output is a distribution summary, specifically summarizing counts into non-overlapping ranges.
+1. SUMMARY: This is a more generic way to request information about a distribution, perhaps represented in some vendor-specific way / not a histogram.
+
+## Trade-offs and mitigations
+
+This avoids requiring programmers to use the `view` API, which is an SDK API, not a user-facing instrumentation API. Letting the application programmer recommend aggregations directly gives the implementation more information about the raw statistics. Letting programmers declare their intent has few downsides, since there is a well-defined default behavior.
+
+## Prior art and alternatives
+
+Existing systems generaly declare separate Metric types according to the desired aggregation.  Raw statistics were invented to overcome this, and the present proposal brings back the ability to specify an Aggregation at the point where a Metric is defined.
+
+## Open questions
+
+There are questions about the value of the MIN and MAX aggregations.  While they are simple to compute, they are difficult to use in practice.
+
+There are questions about the interpretation of HISTOGRAM and SUMMARY. The point of Raw statistics was that we shouldn't specify these aggregations because they are expensive and many implementations are possible.  This is still true. What is the value in specifying HISTOGRAM as opposed to SUMMARY?  How is SUMMARY different from MIN/MAX/COUNT/SUM?
```

Comment:
Not related here but for my understanding
```
class SomeServerImpl(opentelem)
  constructor():
    self.byteCounter = opentelel.NewSumMetric("ingress_bytes")

  handleGRPC(req, res):
    self.byteCounter.Inc(req.ByteSize())
```
Does this report to the sdk that a metric named ingress_bytes has been incremented  or is byteCounter a real object that increments its internal value by 1 and reports its value on a time interval? What if I want to use a special type of aggregation because 1) opentelemetry doesn't support or 2) there's some fidelity that gets lost between the Increment and the exporter? Is this more of an SDK concern?

Reply:
The "heavy" client I'm referring to would go the aggregation itself. A low-level exporter would let the exported-to-library do that itself.